alf/algorithms/sac_algorithm.py

```diff
@@ -423,18 +426,48 @@ def predict_step(self,
             state=state.action,
             epsilon_greedy=epsilon_greedy,
             eps_greedy_sampling=True)
-        empty_state = nest.map_structure(lambda x: (), self.train_state_spec)
-        new_state = empty_state._replace(action=action_state)
         return AlgStep(
             output=action,
-            state=new_state,
+            state=SacState(action=action_state),
             info=SacInfo(action_distribution=action_dist))
 
     def rollout_step(self, time_step: TimeStep, state: SacState):
+        """``rollout_step()`` basically predicts actions like what is done by
+        ``predict_step()``. Additionally, if states are to be stored a in replay
+        buffer, then this function also call ``_critic_networks`` and
+        ``_target_critic_networks`` to maintain their states.
+        """
+        action_dist, action, _, action_state = self._predict_action(
+            time_step.observation,
+            state=state.action,
+            epsilon_greedy=1.0,
+            eps_greedy_sampling=True)
+
         if self.need_full_rollout_state():
-            raise NotImplementedError("Storing RNN state to replay buffer "
-                                      "is not supported by SacAlgorithm")
-        return self.predict_step(time_step, state, epsilon_greedy=1.0)
+            _, critics_state = self._compute_critics(
+                self._critic_networks, time_step.observation, action,
+                state.critic.critics)
+            _, target_critics_state = self._compute_critics(
+                self._target_critic_networks, time_step.observation, action,
+                state.critic.target_critics)
+            critic_state = SacCriticState(
+                critics=critics_state, target_critics=target_critics_state)
+            if self._act_type == ActionType.Continuous:
+                # During unroll, the computations of ``critics_state`` and
+                # ``actor_state`` are the same.
+                actor_state = critics_state
```

Comment:
This doesn't seem correct. SacActionState and SacCriticState are different.

Reply:
This is not ActionState. This is the critics state used for training the actor. In the SAC code, actor_state is used in actor_train_step like below:

```
critics, critics_state = self._compute_critics(
                self._critic_networks, exp.observation, action, state)
```

On the other hand, critics_state is used:

```
critics, critics_state = self._compute_critics(
            self._critic_networks, exp.observation, exp.action, state.critics)
```

They are different in train_step. But in rollout_step, they should be the same.